rewrite-java/src/main/java/org/openrewrite/java/cleanup/CovariantEqualsVisitor.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2020 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.cleanup;
+
+import org.openrewrite.Incubating;
+import org.openrewrite.java.AnnotationMatcher;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Incubating(since = "7.0.0")
+public class CovariantEqualsVisitor<P> extends JavaIsoVisitor<P> {
+
+    private static final MethodMatcher OBJECT_EQUALS_SIGNATURE = new MethodMatcher("* equals(java.lang.Object)");
+
+    public CovariantEqualsVisitor() {
+        setCursoringOn();
+    }
+
+    @Override
+    public J.ClassDecl visitClassDecl(J.ClassDecl classDecl, P p) {
+        J.ClassDecl cd = super.visitClassDecl(classDecl, p);
+        Stream<J.MethodDecl> mds = cd.getBody().getStatements().stream().filter(J.MethodDecl.class::isInstance).map(J.MethodDecl.class::cast);
+        if (mds.noneMatch(m -> OBJECT_EQUALS_SIGNATURE.matches(m, classDecl))) {
+            cd = (J.ClassDecl) new ChangeCovariantEqualsMethodVisitor<>(cd).visit(cd, p, getCursor());
+        }
+        return cd;
+    }
+
+    private static class ChangeCovariantEqualsMethodVisitor<P> extends JavaIsoVisitor<P> {
+
+        private static final AnnotationMatcher OVERRIDE_ANNOTATION_SIGNATURE = new AnnotationMatcher("@java.lang.Override");
+        private static final String EQUALS_BODY_PREFIX_TEMPLATE = "{\n" +
+                "if (#{} == this) return true;\n" +
+                "if (#{} == null || getClass() != #{}.getClass()) return false;\n" +
+                "#{} #{} = (#{}) #{};\n" +
+                "#{}\n" +
+                "}";
```

Comment:
There are other ways of doing this check, by the way. Just using what was already there. Could also do something like

```java
		if (#{} == this) return true;
		if (!(#{} instanceof#{})) return false;
		#{} #{} = (#{}) #{};
...
```

etc.

Reply:
We're using the body that IntelliJ generates for consistency.